s3stream/src/main/java/com/automq/stream/s3/cache/blockcache/AsyncSemaphore.java

```diff
@@ -37,34 +38,37 @@ public AsyncSemaphore(long permits) {
      * @param eventLoop       the eventLoop to run the task when the permits are available
      * @return true if the permits are acquired, false if the task is added to the waiting queue.
      */
-    public synchronized boolean acquire(long requiredPermits, Supplier<CompletableFuture<?>> task,
-        EventLoop eventLoop) {
-        if (permits >= 0) {
-            // allow permits minus to negative
-            permits -= requiredPermits;
-            try {
-                task.get().whenComplete((nil, ex) -> release(requiredPermits));
-            } catch (Throwable e) {
-                LOGGER.error("Error in task", e);
+    public boolean acquire(long requiredPermits, Supplier<CompletableFuture<?>> task,
+                           EventLoop eventLoop) {
+        for (; ; ) {
+            long current = permits.get();
+            if (current <= 0) {
+                tasks.add(new AsyncSemaphoreTask(requiredPermits, task, eventLoop));
+                return false;
+            }
+
+            if (permits.compareAndSet(current, current - requiredPermits)) {
+                try {
+                    task.get().whenComplete((nil, ex) -> release(requiredPermits));
+                } catch (Throwable e) {
+                    LOGGER.error("Error in task", e);
+                    release(requiredPermits); // Release permits on error
+                }
+                return true;
             }
-            return true;
-        } else {
-            tasks.add(new AsyncSemaphoreTask(requiredPermits, task, eventLoop));
-            return false;
         }
     }
 
-    public synchronized boolean requiredRelease() {
-        return permits <= 0 || !tasks.isEmpty();
+    public boolean requiredRelease() {
+        return permits.get() <= 0 || !tasks.isEmpty();
     }
 
-    public synchronized long permits() {
-        return permits;
+    public long permits() {
+        return permits.get();
     }
 
-    synchronized void release(long requiredPermits) {
-        permits += requiredPermits;
-        if (permits > 0) {
+    void release(long requiredPermits) {
```

Comment:
Acquire and release may cause concurrency issues:
1. Thread1: acquire long current = permits.get(), finds it is less than 0
2. Thread2: release permits.addAndGet(requiredPermits) > 0, tries to retrieve a task from the tasks queue, but it is still empty at this point
3. Thread1: places the task into tasks, but no one is awakened at this moment

AsyncSemaphore is only used during cold reading, with low concurrency, temporarily not a performance bottleneck point, can be optimized later.

Reply:
agree. we can do this later : - )